plugins/bil/bil_live.ml

```diff
@@ -0,0 +1,28 @@
+open Core_kernel
+open Bap.Std
+open Bil.Types
+
+let run bil =
+  let (--) = Set.remove in
+  let (++) = Set.union in
+  let free = Exp.free_vars in
+  let is_dead var live =
+    Var.is_virtual var && not (Set.mem live var) in
+  let rec loop init bil =
+    List.fold (List.rev bil) ~init:([], init)
+      ~f:(fun (bil, live) s -> match s with
+          | Special _ | CpuExn _ -> s :: bil, live
+          | Jmp e -> s :: bil, free e ++ live
+          | If (cond, yes, no) ->
+            let yes, live' = loop live yes in
+            let no, live'' = loop live no in
+            let live = free cond ++ live' ++ live'' in
+            Bil.if_ cond yes no :: bil, live
+          | Move (var,e) ->
+            if is_dead var live then bil, live
+            else s :: bil, free e ++ (live -- var)
+          | While (cond, body) ->
```

Comment:
in the body of a while loop the liveness property is propagated in both directions, since the body has to outcoming edges, one that is connected with the rest of the program (and brings liveness from it) and another is from the loop body itself, that brings liveness from all statements of loop body. In your algorithm you effectively ignoring this information, thus you will remove variables whose liveness are guaranteed by preceding statements in the loop body. 
E.g.,

```c
x := 0;
y := 0;
while (i < 2) {
    y = x + 1;
    x = 1;
    i = i + 1;
}
```

assuming that `x` is not live and that `y` is live in the rest of the program, then your analysis will transform the program into
```c
x := 0;
// y := 0; -- killed since `y` is not live here anymore
while (i < 2) {
    y = x + 1;
    // x = 1; --- killed since `x` is dead
    i = i + 1; // this would be also removed, since you didn't treat cond right. 
}
```

To fix, an easy solution would be to stop killing _and_ removing once you enter a while loop. (Remember that this property is transitive to all contents of the while loop, e.g., in `while (c) {x; y; if (d) {p} else {q}; z}`, statements `p` and `q`, as well as `x`, `y`, `z` should be preserved).